src/esp/scene/SceneNode.cpp

```diff
@@ -12,26 +14,173 @@ namespace scene {
+
+  MagnumObject::setParent(newParent);
+
+  // Update new ancestors'SubtreeSensorSuites
+  addSubtreeSensorsToAncestors();
+  // Add sensors to newParent's nodeSensorSuite
+  addSensorToParentNodeSensorSuite();
+  return *this;
+}
+
+void SceneNode::addSensorToParentNodeSensorSuite() {
+  // Only add a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  // Get the first one, and if it is valid, add it to parent's nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    SceneNode* parentNode = dynamic_cast<SceneNode*>(this->parent());
+    // SceneNode::addSensorToParentNodeSensorSuite() is only called when
+    // constructing a sensor or setting the new parent of a node, so parentNode
+    // should never be nullptr
+    CORRADE_ASSERT(parentNode != nullptr,
+                   "SceneNode::addSensorToParentNodeSensorSuite(): Parent node "
+                   "is nullptr", );
+    parentNode->getNodeSensorSuite().add(it->second);
+  }
+}
+
+void SceneNode::removeSensorFromParentNodeSensorSuite() {
+  // Only remove a sensor if this is a leaf node and sensor exists
+  if (!(getSceneNodeTags() & SceneNodeTag::Leaf) ||
+      nodeSensorSuite_->getSensors().empty()) {
+    return;
+  }
+  // There only exists 0 or 1 sensors in a leaf nodeSensorSuite
+  CORRADE_INTERNAL_ASSERT(nodeSensorSuite_->getSensors().size() == 1);
+  // Get the first one, and if it is valid, remove it from parent's
+  // nodeSensorSuite
+  std::map<std::string, std::reference_wrapper<sensor::Sensor>>::iterator it =
+      nodeSensorSuite_->getSensors().begin();
+  if (it != nodeSensorSuite_->getSensors().end()) {
+    SceneNode* parentNode = dynamic_cast<SceneNode*>(this->parent());
+    // If parentNode has not been deconstructed, remove leaf node's sensor from
+    // parentNode's nodeSensorSuite
+    if (parentNode != nullptr) {
+      parentNode->getNodeSensorSuite().remove(it->first);
+    }
+  }
+}
+
+void SceneNode::addSubtreeSensorsToAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
+      currentNode->getSubtreeSensorSuite().add(sensor.second);
+      currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
+    }
+  }
+}
+
+void SceneNode::removeSubtreeSensorsFromAncestors() {
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = dynamic_cast<SceneNode*>(this->parent());
+    while (currentNode && !SceneGraph::isRootNode(*currentNode)) {
```

Comment:
if `currentNode` is the root node, then this piece of code will not remove the sensors from the `currentNode`.

Reply:
Go back to the slack channel, read the code I gave you very carefully, and try to understand the difference between my code and the code you wrote here.

Reply:
In your tests, DO check if the number of sensors is correct in the root node!!

Reply:
Thank you, this is really important. Here is the code from slack:
```
 SceneNode* currentNode = this;
  do {
    currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
    // no need to worry currentNode could be nullptr, no chance
    currentNode->getSubtreeSensorSuite().getSensors().erase(sensor.first);
  } while (!SceneGraph::isRootNode(*currentNode));
  
```
We had discussed previously that currentNode could never be nullptr, which made sense because the currentNode would never be nullptr before reaching rootNode's parent, which it would not reach due to the while condition. However, this currentNode was actually coming back nullptr sometimes, and this was causing the segfault when running examples.py. My update to check that currentNode was not nullptr fixed the segfault, but overlooked adding/removing sensors from the root node's SensorSuites.

I have updated the logic to check that currentNode is not nullptr and to also update rootNode. I have also added checks for rootNode's SensorSuites in SensorTest.

Thanks and let me know if you have any questions.